File: LeetCode Mastery/1. Two Sum/two-sum/src/Solution_Two.java
import java.util.Arrays;
import java.util.HashMap;

/*
 * Solution 2
 * - double pass HashMap
 * - O(n^2) but better access time
 * - runtime: 4 ms
 * - memory: 44.61 MB
 */
public class Solution_Two {
    public Solution_Two(int[] input_nums, int input_target) {
        System.out.println("\nSolution 2 -->");
        System.out.println("nums[]: " + Arrays.toString(input_nums));
        System.out.println("target: " + input_target);
        System.out.println("output: " + Arrays.toString(twoSum(input_nums, input_target)));
    }
    
    private int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> nums_map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            nums_map.put(nums[i], i);
        }
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (nums_map.containsKey(complement) && (nums_map.get(complement) != i)) {
                return new int[] {i, nums_map.get(complement)};
            }
        }
        return null;
    }
}

File: LeetCode Mastery/1. Two Sum/two-sum/src/Solution_Two.py
# Solution 2
# - double pass HashMap
# - O(n^2) but better access time
# - runtime: 58 ms
# - memory: 17.69 MB

from typing import List

class Solution_Two:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        nums_map = {}
        for i in range(n):
            nums_map[nums[i]] = i
        for i in range(n):
            complement = target - nums[i]
            if complement in nums_map and nums_map[complement] != i:
                return [i, nums_map[complement]]

    nums = [2, 7, 11, 15]
    target = 9
    print(twoSum(self = True, nums = nums, target = target))